From: Daniel Mack <daniel@zonque.org>
Date: Wed, 3 Jun 2015 13:33:26 +0200
Subject: util: fix another cunescape() regression

Fix a regression caused by 4034a06d ("util: rework word parsing and c
unescaping code") which broke octal escape sequences.

The reason for this breakage is that cunescape_one() expects 4 characters
in an octal encoding, which is a stray left-over from the old code which
operated on different variables to make the length check.

While at it, add a test case to prevent the same thing from happening
again.
---
 src/shared/util.c    | 2 +-
 src/test/test-util.c | 3 +++
 2 files changed, 4 insertions(+), 1 deletion(-)

diff --git a/src/shared/util.c b/src/shared/util.c
index 34024ba..da7f597 100644
--- a/src/shared/util.c
+++ b/src/shared/util.c
@@ -1186,7 +1186,7 @@ static int cunescape_one(const char *p, size_t length, char *ret, uint32_t *ret_
                 int a, b, c;
                 uint32_t m;
 
-                if (length != (size_t) -1 && length < 4)
+                if (length != (size_t) -1 && length < 3)
                         return -EINVAL;
 
                 a = unoctchar(p[0]);
diff --git a/src/test/test-util.c b/src/test/test-util.c
index 36773c1..fab485f 100644
--- a/src/test/test-util.c
+++ b/src/test/test-util.c
@@ -459,6 +459,9 @@ static void test_cunescape(void) {
         assert_se(cunescape("\\u0000", 0, &unescaped) < 0);
         assert_se(cunescape("\\u00DF\\U000000df\\u03a0\\U00000041", UNESCAPE_RELAX, &unescaped) >= 0);
         assert_se(streq_ptr(unescaped, "ßßΠA"));
+
+        assert_se(cunescape("\\073", 0, &unescaped) >= 0);
+        assert_se(streq_ptr(unescaped, ";"));
 }
 
 static void test_foreach_word(void) {
